kiwi/lib/search.py

```diff
@@ -0,0 +1,438 @@
+            )
+            kiwi_config['system']['model']['outputs'][
+                'sentence_loss_weight'
+            ] = sentence_loss_weight
+            names.append('sentence_loss_weight')
+            values.append(sentence_loss_weight)
+
+    specified_word_level = kiwi_config['system']['model']['outputs']['word_level'][
+        'target'
+    ]
+    if specified_word_level and config.options.search_hter:
+        if hter and config.options.search_sentence_loss_weight:
+            # Also search for the sentence weight
+            sentence_loss_weight = trial.suggest_uniform(
+                'sentence_loss_weight',
+                config.options.sentence_loss_weight.lower,
+                config.options.sentence_loss_weight.upper,
+            )
+            kiwi_config['system']['model']['outputs'][
+                'sentence_loss_weight'
+            ] = sentence_loss_weight
+            names.append('sentence_loss_weight')
+            values.append(sentence_loss_weight)
+
+    specified_sentence_level = kiwi_config["system"]["model"]["outputs"][
+        "sentence_level"
+    ]["hter"]
+    if specified_sentence_level and config.options.search_sentence_loss_weight:
+        # Also search for the sentence weight
+        sentence_loss_weight = trial.suggest_uniform(
+            'sentence_loss_weight',
+            config.options.sentence_loss_weight.lower,
+            config.options.sentence_loss_weight.upper,
+        )
+        kiwi_config["system"]["model"]["outputs"][
+            "sentence_loss_weight"
+        ] = sentence_loss_weight
+        names.append("sentence_loss_weight")
+        values.append(sentence_loss_weight)
+
+    if kiwi_config['quiet'] is None:
+        kiwi_config['quiet'] = False
+    if kiwi_config['verbose'] is None:
+        kiwi_config['verbose'] = False
+    trainconfig = train.Configuration(**kiwi_config)
+
+    # TODO: integrate PTL trainer callback
+    # ptl_callback = PyTorchLightningPruningCallback(trial, monitor='val_PEARSON')
+    # train_info = train.run(trainconfig, ptl_callback=ptl_callback)
+
+    logger.info(f"############# STARTING TRIAL {trial.number} #############")
+    logger.info(f"PARAMETERS: {dict(zip(names, values))}")
+    for name, value in zip(names, values):
+        logger.info(f'{name}: {value}')
+
+    try:
+        train_info = train.run(trainconfig)
+    except RuntimeError as e:
+        logger.info(f'ERROR OCCURED; SKIPPING TRIAL: {e}')
+        return -1
+
+    logger.info(f"############# TRIAL {trial.number} FINISHED #############")
+    result = train_info.best_metrics.get(main_metric, -1)
+    if result == -1:
+        logger.error(
+            'Trial did not converge:\n'
+            f'    `train_info.best_metrics={train_info.best_metrics}`\n'
+            f'Setting result to -1'
+        )
+
+    logger.info(f"RESULTS: {result} {main_metric}")
+    logger.info(f"PARAMETERS: {dict(zip(names, values))}")
```

Comment:
isn't this and the for-loop below printing the same thing?

Reply:
Yes, but one in a one-liner east to find with grep 😅  😬

Reply:
And the other more pleasing to the eye

Reply:
Can I have both, pleease?

Reply:
Sure, this doesn't seem like a big detail 🙂 

(In my personal scripts I also do it this way)